wiki_articles/as_if.md

```diff
@@ -0,0 +1,41 @@
+# What Is the As-If Rule, and How Does the Compiler Optimize Code?
+
+The compiler can transform code however it thinks is best, as long as this
+doesn't change the *observable behavior* of the program.
+People call this the *[as-if rule][as-if]*.
+Observable behavior includes, but is not limited to:
+- access of **[volatile][volatile]** objects
+- writing output to the terminal (**[printf][printf]**, **[std::cout][cout]**, etc.)
+- aborting the program (**[exit][exit]**, **[std::terminate][term]**, etc.)
+
+<!-- inline -->
+## Example
+```cpp
+int main() {
+  if (false)
+    return 1;
+  int x = 3;
+  return x + 7;
+}
+```
+The *observable behavior* is that the program returns the exit code `10`.
+
+<!-- inline -->
+## Assembly Output
+```x86asm
+main:
+; 'if (false)' removed by
+; dead-code elimination.
+; 'x' is constant-folded.
```

Comment:
Super minor but what's happening in this case is usually called constant-propagation, but that likely won't fit on one line.

Reply:
It wouldn't fit on one line, so I think it should just stay the way it is. People colloquially refer to constant-propagation by constant-folding anyway, even if it's not strictly correct. I think that's okay.